### PR TITLE
Local Model Node CR

### PR DESCRIPTION
**Pull request by HotsauceLee**
**What this PR does / why we need it**:

This is the continuation of  #3860 

In current implementation of model cache, ClusterLocalModel controller owns the download jobs that download models to local nodes, yet there's no daemonset watching the state of those local models. This PR is the first out of 3 steps to move ownership of local node to a new LocalModelNode controller. All 3 steps are:

- Create LocalModelNode CR (this PR)
- Create LocalModelNode daemonset controller
- Update ClusterLocalModel controller

Why do we need the daemonset contoller?

1. Re-download the model if it is missing from the disk
2. Make deletion easier. Currently we have a job to delete models when the local model CR is removed. However, node failure (not ready) would prevent the CR from being deleted since the job would not ran. The daemonset controller can delete the model on its own without creating a job, thus non-ready nodes would not block deletion of local model CRs.
3. Separation of control to make the code cleaner

image

Example LocalModelNode CR:

```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LocalModelNode
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.kserve.io/v1alpha1","kind":"LocalModelNode","metadata":{"annotations":{},"name":"local-model-node-test"},"spec":{"localModels":[{"modelName":"test-model1","sourceModelUri":"s3://path/to/test-model1"},{"modelName":"test-model2","sourceModelUri":"s3://path/to/test-model2"}]}}
  creationTimestamp: "2024-10-10T20:57:51Z"
  generation: 1
  name: local-model-node-test
  resourceVersion: "1660862"
  uid: 87a41d6a-3951-4107-830d-0d809ff88b9d
spec:
  localModels:
    - modelName: test-model1
      sourceModelUri: s3://path/to/test-model1
    - modelName: test-model2
      sourceModelUri: s3://path/to/test-model2
```

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

Applied the following CRD:
```yaml
---
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.16.2
  name: localmodelnodes.serving.kserve.io
spec:
  group: serving.kserve.io
  names:
    kind: LocalModelNode
    listKind: LocalModelNodeList
    plural: localmodelnodes
    singular: localmodelnode
  scope: Cluster
  versions:
  - name: v1alpha1
    schema:
      openAPIV3Schema:
        properties:
          apiVersion:
            type: string
          kind:
            type: string
          metadata:
            type: object
          spec:
            properties:
              localModels:
                items:
                  properties:
                    modelName:
                      type: string
                    sourceModelUri:
                      type: string
                  required:
                  - modelName
                  - sourceModelUri
                  type: object
                type: array
            required:
            - localModels
            type: object
          status:
            properties:
              modelStatus:
                additionalProperties:
                  enum:
                  - ""
                  - ModelDownloadPending
                  - ModelDownloading
                  - ModelDownloaded
                  - ModelDownloadError
                  type: string
                type: object
            type: object
        type: object
    served: true
    storage: true
    subresources:
      status: {}
```

Then applied the following CR:
```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LocalModelNode
metadata:
  name: local-model-node-test
spec:
  localModels:
    - modelName: test-model1
      sourceModelUri: s3://path/to/test-model1
    - modelName: test-model2
      sourceModelUri: s3://path/to/test-model2
```

Result:
```
yli1109@RVQWYXM093 kserve % k get localmodelnode local-model-node-test -n kserve
NAME                    AGE
local-model-node-test   20s
yli1109@RVQWYXM093 kserve % k get localmodelnode local-model-node-test -n kserve -oyaml | yq
apiVersion: serving.kserve.io/v1alpha1
kind: LocalModelNode
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.kserve.io/v1alpha1","kind":"LocalModelNode","metadata":{"annotations":{},"name":"local-model-node-test"},"spec":{"localModels":[{"modelName":"test-model1","sourceModelUri":"s3://path/to/test-model1"},{"modelName":"test-model2","sourceModelUri":"s3://path/to/test-model2"}]}}
  creationTimestamp: "2024-10-10T20:57:51Z"
  generation: 1
  name: local-model-node-test
  resourceVersion: "1660862"
  uid: 87a41d6a-3951-4107-830d-0d809ff88b9d
spec:
  localModels:
    - modelName: test-model1
      sourceModelUri: s3://path/to/test-model1
    - modelName: test-model2
      sourceModelUri: s3://path/to/test-model2

```

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: pkg/apis/serving/v1alpha1/local_model_node_types.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type SourceDestination struct {
+	SourceURI   string `json:"sourceUri"`
+	Destination string `json:"destination"`
+}
+
+// +k8s:openapi-gen=true
+type LocalModelNodeSpec struct {
+	SourceDestinations []SourceDestination `json:"sourceDestinations"`
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/local_model_node_types.go]: Can you add comments to explain the usage of this field?

File: pkg/apis/serving/v1alpha1/local_model_node_types.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type SourceDestination struct {
+	SourceURI   string `json:"sourceUri"`
+	Destination string `json:"destination"`
+}
+
+// +k8s:openapi-gen=true
+type LocalModelNodeSpec struct {
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/local_model_node_types.go:27]: I think these fields should all be part of the LocalNodeNodeStatus

Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/local_model_node_types.go:29]: > I think these fields should all be part of the LocalModelNodeStatus. Initially the list is empty and the status is updated when models are cached on the local node.

File: pkg/apis/serving/v1alpha1/local_model_node_types.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	ModelStatus        *ModelStatus        `json:"modelStatus,omitempty"`
+}
+
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Cluster"
+type LocalModelNode struct {
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/local_model_node_types.go:40]: Refer to design of Kubernetes node
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.31/#node-v1-core

File: pkg/apis/serving/v1alpha1/local_model_node_types.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type SourceDestination struct {
+	SourceURI   string `json:"sourceUri"`
```
Comment (greenmoon55) [inline pkg/apis/serving/v1alpha1/local_model_node_types.go]: Can you add comments for all fields in this file?

File: pkg/apis/serving/v1alpha1/local_model_node_types.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	Destination string `json:"destination"`
```
Comment (greenmoon55) [inline pkg/apis/serving/v1alpha1/local_model_node_types.go]: Rename to Model name? folder name?

File: pkg/apis/serving/v1alpha1/local_model_node_types.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type SourceDestination struct {
```
Comment (greenmoon55) [inline pkg/apis/serving/v1alpha1/local_model_node_types.go]: Rename - ModelInfo?

File: pkg/apis/serving/v1alpha1/local_model_node_types.go (hunk 2 above)
Comment (greenmoon55) [reply pkg/apis/serving/v1alpha1/local_model_node_types.go:29]: I was thinking `SourceDestinations` is in the spec which is generated by the local model controller  and `ModelStatus` should be in the status.

**Review by greenmoon55 — COMMENTED** [review]
Thanks! Just some minor comments! Can you test it by applying the CRDs to your cluster and then create a resource?

**Review by greenmoon55 — COMMENTED** [review]
Can you add a bit more details to the description of this pr?

File: pkg/apis/serving/v1alpha1/local_model_node_types.go (hunk 3 above)
Comment (HotsauceLee) [reply pkg/apis/serving/v1alpha1/local_model_node_types.go:40]: moved model status to LocalModelNodeStatus

**Comment by greenmoon55** [comment]
Can you add these to the description?

> Why do we need the daemonset contoller?
> 1. Re-download the model if it is missing from the disk
> 2. Make deletion easier. Currently we have a job to delete models when the local model CR is removed. However, node failure (not ready) would prevent the CR from being deleted since the job would not ran. The daemonset controller can delete the model on its own without creating a job, thus non-ready nodes would not block deletion of local model CRs. 
> 4. Separation of control to make the code cleaner

**Review by greenmoon55 — APPROVED** [review]

**Review by yuzisun — APPROVED** [review]
/lgtm
/approve

**Pull request merged**